---
title: Processing the Rizzardi/Hickey brain dataset
author: Peter Hickey
date: 17 August 2020
output:
  BiocStyle::html_document:
    titlecaps: false
    toc_float: true
bibliography: ../ref.bib
editor_options: 
  chunk_output_type: console
---

```{r style, echo = FALSE, results = "hide", message = FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error = FALSE, message = FALSE, warning = FALSE)
```

# Download the data

We obtain a whole-genome bisulfite-sequencing dataset of human brain from @rizzardi2019neuronal.
Count matrices of methylated reads and total reads for various samples, cytosine contexts (CpG, CpA, CpT), and strands (unstranded, forward, reverse) are provided from the Gene Expression Omnibus using accession code [GSE96612](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96612).
We download these using `r Biocpkg("BiocFileCache")` to cache the results:

```{r}
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask = FALSE)

base_url <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE96nnn/GSE96612/suppl"

cov_bulk_cpg_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_Cov_matrix.bulk.CpG_unstranded.txt.gz"))
m_bulk_cpg_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_M_matrix.bulk.CpG_unstranded.txt.gz"))

cov_sorted_cpg_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_Cov_matrix.sorted.CpG_unstranded.txt.gz"))
m_sorted_cpg_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_M_matrix.sorted.CpG_unstranded.txt.gz"))

cov_sorted_cpa_fwd_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_Cov_matrix.sorted.CpA_forward_strand.txt.gz"))
m_sorted_cpa_fwd_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_M_matrix.sorted.CpA_forward_strand.txt.gz"))

cov_sorted_cpa_rev_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_Cov_matrix.sorted.CpA_reverse_strand.txt.gz"))
m_sorted_cpa_rev_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_M_matrix.sorted.CpA_reverse_strand.txt.gz"))

cov_sorted_cpt_fwd_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_Cov_matrix.sorted.CpT_forward_strand.txt.gz"))
m_sorted_cpt_fwd_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_M_matrix.sorted.CpT_forward_strand.txt.gz"))

cov_sorted_cpt_rev_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_Cov_matrix.sorted.CpT_reverse_strand.txt.gz"))
m_sorted_cpt_rev_fname <- bfcrpath(
  bfc,
  file.path(base_url, "GSE96612_M_matrix.sorted.CpT_reverse_strand.txt.gz"))
```

# Process the data and save for upload

We create a function to extract the loci, sample-level metadata, and count matrices from each file and write all of the relevant components to to a [HDF5 file](https://en.wikipedia.org/wiki/Hierarchical_Data_Format) for upload to `r Biocpkg("ExperimentHub")`.

```{r}
library(data.table)
library(GenomicRanges)
library(HDF5Array)

FUN <- function(cov_fname, m_fname, out_fname) {
  # Load data
  cov_dt <- fread(cov_fname)
  m_dt <- fread(m_fname)
  stopifnot(
    identical(colnames(cov_dt), colnames(m_dt)),
    identical(cov_dt$chr, m_dt$chr),
    identical(cov_dt$start, m_dt$start),
    identical(cov_dt$strand, m_dt$strand))
  
  # rowRanges
  rowranges <- GPos(
    seqnames = cov_dt$chr, 
    pos = cov_dt$start, 
    strand = cov_dt$strand, 
    seqinfo = Seqinfo(genome = "hg19"))
  stopifnot(!is.unsorted(rowranges))
  h5write(
    # TODO: https://github.com/Bioconductor/GenomicRanges/issues/46
    obj = as.data.frame(rowranges), 
    file = out_fname, 
    name = "rowRanges",
    level = level,
    # NOTE: Needed to ensure factor columns aren't converted to their underlying 
    #       integer types.
    DataFrameAsCompound = FALSE)
  h5write(
    obj = as.data.frame(seqinfo(rowranges)), 
    file = out_fname, 
    name = "rowRanges/seqinfo",
    level = level,
    # NOTE: Needed to ensure factor columns aren't converted to their underlying 
    #       integer types.
    DataFrameAsCompound = FALSE)
  
  # colData
  cns <- setdiff(colnames(cov_dt), c("chr", "start", "strand"))
  cns <- sub("NA$", "NAcc", cns)
  cn_split <- strsplit(cns, "_")
  donor <- vapply(cn_split, "[[", 1, FUN.VALUE = character(1L))
  tissue <- vapply(cn_split, "[[", 2, FUN.VALUE = character(1L))
  neun <- if (all(lengths(cn_split)) == 3) {
    vapply(cn_split, "[[", 2, FUN.VALUE = character(1L))
  } else {
    "unsorted"
  }
  coldata <- data.frame(donor = donor, tissue = tissue, neun = neun)
  h5write(
    obj = coldata, 
    file = out_fname, 
    name = "colData",
    level = level,
    # NOTE: Needed to ensure factor columns aren't converted to their underlying 
    #       integer types.
    DataFrameAsCompound = FALSE)

  # M and Cov
  # TODO: Think about best chunkdim. Default is getHDF5DumpChunkDim(dim(x)),
  #       which currently results in blocks along columns.
  chunkdim <- makeCappedVolumeBox(
    getHDF5DumpChunkLength(),
    dim(M),
    "hypercube")
  M <- DelayedArray(m_dt[, !c("chr", "start", "strand")])
  type(M) <- "integer"
  rownames(M) <- NULL
  colnames(M) <- sub("NA$", "NAcc", colnames(M))
  stopifnot(identical(cns, colnames(M)))
  M <- writeHDF5Array(
    x = M,
    filepath = out_fname,
    name = "M",
    chunkdim = chunkdim,
    level = level,
    with.dimnames = TRUE)
  Cov <- DelayedArray(cov_dt[, !c("chr", "start", "strand")])
  type(Cov) <- "integer"
  rownames(Cov) <- NULL
  colnames(Cov) <- sub("NA$", "NAcc", colnames(Cov))
  stopifnot(identical(cns, colnames(Cov)))
  Cov <- writeHDF5Array(
    x = Cov,
    filepath = out_fname,
    name = "Cov",
    chunkdim = chunkdim,
    level = level,
    with.dimnames = TRUE)

  list(M = M, Cov = Cov, rowRanges = rowranges, colData = coldata)
}
```

We run this on all the datasets.

```{r}
options(DelayedArray.block.size = 1e10) # 10GB block size.
path <- file.path("MethylSeqData", "rizzardi_hickey_brain")
dir.create(path, showWarnings = FALSE, recursive = TRUE)
level <- 9

bulk_cpg <- FUN(
  cov_bulk_cpg_fname, 
  m_bulk_cpg_fname, 
  file.path(path, "bulk_CpG.h5"))
sorted_cpg <- FUN(
  cov_sorted_cpg_fname, 
  m_sorted_cpg_fname, 
  file.path(path, "sorted_CpG.h5"))
sorted_cpa_fwd <- FUN(
  cov_sorted_cpa_fwd_fname, 
  m_sorted_cpa_fwd_fname, 
  file.path(path, "sorted_CpA_fwd.h5"))
sorted_cpa_rev <- FUN(
  cov_sorted_cpa_rev_fname, 
  m_sorted_cpa_rev_fname, 
  file.path(path, "sorted_CpA_rev.h5"))
sorted_cpt_fwd <- FUN(
  cov_sorted_cpt_fwd_fname, 
  m_sorted_cpt_fwd_fname, 
  file.path(path, "sorted_CpT_fwd.h5"))
sorted_cpt_rev <- FUN(
  cov_sorted_cpt_rev_fname, 
  m_sorted_cpt_rev_fname, 
  file.path(path, "sorted_CpT_rev.h5"))
```

# Session info

```{r}
sessionInfo()
```

# References
